File: planctl/invocation.py
"""Invocation payload builder for planctl CLI invocations.

Mutating verbs call ``build_planctl_invocation(verb, target_id, ...)`` to build
the ``planctl_invocation`` envelope payload.  The hook registered in
``apps/hookctl/hooks/planctl-mutation/`` consumes this payload and commits the
``.planctl/`` state — the CLI itself no longer shells out to git.

Read-only verbs (and the click decorator path) call
``build_planctl_invocation_readonly(verb, target_id, ...)`` to build a
lighter-weight payload with NULL ``subject``, ``files``, and ``commit_sha``.
The hook INSERTs a row and emits to UDS but skips the git commit step.

Commit message format (produced by the hook for mutating verbs)::

    chore(planctl): <verb> <id>[ — <detail>]

    Planctl-Op: <verb>
    Planctl-Target: <id>
    Planctl-Prev-Op: <sha of HEAD captured BEFORE git commit>
    Session-Id: <CLAUDE_CODE_SESSION_ID>   (omitted when the env var is absent)

Envelope fields (mutating verbs):
    repo_root   — the cwd-derived repo at invocation time (always present).
    state_repo  — the repo whose .planctl/ carries state for this epic/task.
                  Populated from epic.primary_repo when the verb targets an
                  epic/task; falls back to repo_root for cwd-only verbs (init,
                  detect).  The hook uses state_repo when present to route the
                  commit to the correct .planctl/ directory.  Both fields are
                  always emitted; they are equal for single-repo projects.
"""

from __future__ import annotations

import os
from pathlib import Path


def _planctl_dir(repo_root: Path) -> Path:
    """Resolve the .planctl/ directory relative to repo_root."""
    return repo_root / ".planctl"


def build_planctl_invocation(
    verb: str,
    target_id: str,
    detail: str | None = None,
    *,
    repo_root: Path,
    primary_repo: str | None = None,
    queue_jump: bool = False,
) -> dict:
    """Build the ``planctl_invocation`` envelope payload for a mutating verb.

    This payload is injected into the JSON envelope by ``output.emit()`` so
    that downstream hooks can parse it and commit the ``.planctl/`` state
    without the CLI needing to shell out to git itself.

    Parameters
    ----------
    verb:
        Canonical verb name registered in ``commit_messages.VERB_TEMPLATES``.
    target_id:
        Epic or task ID that was mutated.
    detail:
        Optional short detail appended after em-dash in the subject.
    repo_root:
        Project root (directory containing ``.planctl/`` and ``.git``).
        Semantics: the cwd-derived repo at invocation time.
    primary_repo:
        The repo whose ``.planctl/`` carries the state for this epic/task.
        Populated from ``epic.primary_repo`` for verbs targeting an epic/task;
        defaults to ``None`` (falls back to ``repo_root`` in the envelope).
    queue_jump:
        Whether the epic requested a priority jump. Two sources set
        True: a ``scaffold`` whose epic YAML opts in (``queue_jump: true``),
        and the ``epic queue-jump`` verb (``/plan:next``) flipping an existing
        epic. The field is server-derived from the epic record — keeperd lifts
        it off this envelope into ``epics.queue_jump`` and derives the
        ``!``-prefixed ``sort_path``. Old envelopes lacking the key fold to
        False deterministically (``?? false`` on the keeper side).

    Returns
    -------
    dict
        Payload with keys: ``files``, ``op``, ``target``, ``subject``,
        ``touched_path_files``, ``repo_root``, ``state_repo``, ``queue_jump``,
        ``session_id``.
    """
    from planctl.commit_messages import build_subject
    from planctl.store import _read_touched_files

    planctl_dir = _planctl_dir(repo_root)

    # Resolve session id — CLAUDE_CODE_SESSION_ID is the sole source. Fail
    # closed on None: the claude binary ships this intrinsically on every
    # session (including resumed ones), so it is always present for a planctl
    # mutating verb invocation inside a Claude harness.
    session_id = os.environ.get("CLAUDE_CODE_SESSION_ID") or None
    if session_id is None:
        raise RuntimeError(
            "planctl build_planctl_invocation requires a resolvable session_id; "
            "CLAUDE_CODE_SESSION_ID must be set (the claude binary ships it "
            "intrinsically inside a Claude harness; tests and manual invocations "
            "must set it themselves)."
        )

    # Touched files record paths (for hook cleanup, G7).
    touched_path_files = _read_touched_files(planctl_dir, session_id)

    # Touched paths → intersect with dirty set.
    touched = _read_touched_paths(repo_root, session_id)
    dirty = _dirty_planctl_paths(repo_root)
    files = sorted(p for p in touched if p in dirty)

    subject = build_subject(verb, target_id, detail)

    # state_repo: where .planctl/ state lives.
    # primary_repo (from epic.primary_repo) takes precedence; falls back to
    # repo_root for verbs that have no epic/task target.
    state_repo = primary_repo if primary_repo is not None else str(repo_root)

    return {
        "files": files,
        "op": verb,
        "target": target_id,
        "subject": subject,
        "touched_path_files": touched_path_files,
        "repo_root": str(repo_root),
        "state_repo": state_repo,
        "queue_jump": queue_jump,
        # The resolved committing session id, threaded into the
        # `chore(planctl)` commit's `Session-Id` trailer by
        # `auto_commit_from_invocation`. Same opaque v4 UUID the keeper hook
        # uses (job_id === session_id). Always populated here — the resolve
        # above fails closed for mutating verbs.
        "session_id": session_id,
    }


def _read_touched_paths(repo_root: Path, session_id: str) -> list[str]:
    """Return paths recorded for *session_id* in the touched-paths log.

    Each path is a POSIX string relative to *repo_root*, starting with
    ``.planctl/``.  Paths that fail validation (``..``, non-``.planctl/``
    prefix) are rejected with a ``RuntimeError`` rather than silently dropped
    — a bad path means a bug upstream and should be loud.
    """
    planctl_dir = _planctl_dir(repo_root)
    touched_dir = planctl_dir / "state" / "sessions" / session_id / "touched"
    if not touched_dir.exists():
        return []

    paths: list[str] = []
    for txt in touched_dir.glob("*.txt"):
        raw = txt.read_text(encoding="utf-8").strip()
        if not raw:
            continue
        # Security: reject traversal and non-.planctl/ paths.
        if ".." in raw.split("/"):
            raise RuntimeError(
                f"Touched-paths record contains path traversal: {raw!r} "
                f"(file: {txt}). This is a bug — report it."
            )
        if not raw.startswith(".planctl/"):
            raise RuntimeError(
                f"Touched-paths record contains non-.planctl/ path: {raw!r} "
                f"(file: {txt}). This is a bug — report it."
            )
        paths.append(raw)
    return paths


def build_planctl_invocation_readonly(
    verb: str,
    target_id: str | None = None,
    *,
    repo_root: Path,
) -> dict:
    """Build a read-only ``planctl_invocation`` envelope payload.

    Used by the click decorator for verbs that don't mutate ``.planctl/``
    state (show, cat, validate, list, etc.).  Skips the touched-paths log read
    and git dirty-set intersection — read-only verbs touch nothing.

    The hook INSERTs a row with NULL ``subject``, ``files``, and ``commit_sha``
    and emits an audit row, but skips the git commit step.

    Parameters
    ----------
    verb:
        Canonical verb name (e.g. ``"show"``, ``"cat"``, ``"list"``).
    target_id:
        Epic or task ID, or None when the verb has no target (e.g. ``"list"``).
    repo_root:
        Project root (directory containing ``.planctl/`` and ``.git``).

    Returns
    -------
    dict
        Payload with: ``op``, ``target``, ``subject=None``, ``files=None``,
        ``touched_path_files=[]``, ``repo_root``, ``state_repo``.
    """
    repo_root_str = str(repo_root)
    return {
        "files": None,
        "op": verb,
        "target": target_id,
        "subject": None,
        "touched_path_files": [],
        "repo_root": repo_root_str,
        "state_repo": repo_root_str,
    }


def _dirty_planctl_paths(repo_root: Path) -> set[str]:
    """Return the set of dirty (modified/untracked) .planctl/ paths from git.

    Uses ``--untracked-files=all`` so new files appear individually rather than
    as directory-level ``??`` entries — without this, freshly-written files that
    have never been committed show up as ``?? .planctl/epics/`` instead of
    ``?? .planctl/epics/fn-1-test.json`` and the touched-path intersection
    returns an empty set.
    """
    import subprocess

    result = subprocess.run(
        ["git", "status", "--porcelain", "--untracked-files=all", "--", ".planctl/"],
        cwd=str(repo_root),
        capture_output=True,
        text=True,
    )
    paths: set[str] = set()
    for line in result.stdout.splitlines():
        if len(line) < 4:
            continue
        # git status --porcelain: XY <path> (first 3 chars are status + space)
        rel = line[3:].strip()
        # Handle renames: "old -> new" format
        if " -> " in rel:
            rel = rel.split(" -> ", 1)[1]
        if rel:
            paths.add(rel)
    return paths

File: src/invocation.ts
// Invocation payload builders — the port of planctl/invocation.py.
//
// Field order is the wire contract and must match Python exactly:
//   files, op, target, subject, touched_path_files, repo_root, state_repo
//   (mutating adds queue_jump, session_id after state_repo).
//
// buildPlanctlInvocationReadonly: read-only verbs touch nothing, so files/subject
// are null and touched_path_files is empty; repo_root === state_repo.
//
// buildPlanctlInvocation (mutating): session id fail-CLOSED (throws when
// CLAUDE_CODE_SESSION_ID is absent); files = the sorted intersection of the
// session's touched-paths log with git's dirty .planctl/ set; subject from
// buildSubject; queue_jump + session_id ride after state_repo.

import { existsSync, readdirSync, readFileSync } from "node:fs";
import { join } from "node:path";

import { buildSubject } from "./commit.ts";

export interface ReadonlyInvocation {
  files: null;
  op: string;
  target: string | null;
  subject: null;
  touched_path_files: never[];
  repo_root: string;
  state_repo: string;
}

export function buildPlanctlInvocationReadonly(
  verb: string,
  repoRoot: string,
  target: string | null = null,
): ReadonlyInvocation {
  return {
    files: null,
    op: verb,
    target,
    subject: null,
    touched_path_files: [],
    repo_root: repoRoot,
    state_repo: repoRoot,
  };
}

export interface MutatingInvocation {
  files: string[];
  op: string;
  target: string;
  subject: string;
  touched_path_files: string[];
  repo_root: string;
  state_repo: string;
  queue_jump: boolean;
  session_id: string;
}

/** Build the mutating planctl_invocation payload. Mirrors
 * planctl.invocation.build_planctl_invocation:
 *  - session id from CLAUDE_CODE_SESSION_ID, fail-CLOSED (throw when absent),
 *  - touched_path_files = the session's recorded touched-log filenames,
 *  - files = sorted(touched-paths content ∩ dirty .planctl/ set),
 *  - state_repo = primaryRepo when given, else repoRoot.
 *
 * `target` is the epic/task/project id; `detail` is the optional em-dash
 * subject suffix. Throws RuntimeError-equivalent when the session id is absent
 * (Python raises RuntimeError; here a plain Error with the same message). */
export function buildPlanctlInvocation(
  verb: string,
  target: string,
  detail: string | null | undefined,
  opts: {
    repoRoot: string;
    primaryRepo?: string | null;
    queueJump?: boolean;
  },
): MutatingInvocation {
  const { repoRoot, primaryRepo = null, queueJump = false } = opts;
  const planctlDir = join(repoRoot, ".planctl");

  // Session id — CLAUDE_CODE_SESSION_ID is the sole source. Fail-closed on
  // absence: the claude binary ships it intrinsically on every session, so it is
  // always present for a mutating verb inside a Claude harness.
  const sessionId = process.env.CLAUDE_CODE_SESSION_ID || null;
  if (sessionId === null) {
    throw new Error(
      "planctl build_planctl_invocation requires a resolvable session_id; " +
        "CLAUDE_CODE_SESSION_ID must be set (the claude binary ships it " +
        "intrinsically inside a Claude harness; tests and manual invocations " +
        "must set it themselves).",
    );
  }

  const touchedPathFiles = readTouchedFiles(planctlDir, sessionId);

  const touched = readTouchedPaths(repoRoot, sessionId);
  const dirty = dirtyPlanctlPaths(repoRoot);
  const files = [...touched].filter((p) => dirty.has(p)).sort();

  const subject = buildSubject(verb, target, detail);

  // state_repo: primaryRepo (from epic.primary_repo) takes precedence; falls
  // back to repoRoot for verbs with no epic/task target.
  const stateRepo = primaryRepo !== null ? primaryRepo : repoRoot;

  return {
    files,
    op: verb,
    target,
    subject,
    touched_path_files: touchedPathFiles,
    repo_root: repoRoot,
    state_repo: stateRepo,
    queue_jump: queueJump,
    session_id: sessionId,
  };
}

/** The touched-log record filenames for `sessionId` (for hook cleanup, G7) —
 * the basenames of `*.txt` under sessions/<sid>/touched/. Mirrors
 * store._read_touched_files. Empty list when the dir is absent. */
function readTouchedFiles(planctlDir: string, sessionId: string): string[] {
  const touchedDir = join(
    planctlDir,
    "state",
    "sessions",
    sessionId,
    "touched",
  );
  if (!existsSync(touchedDir)) {
    return [];
  }
  return readdirSync(touchedDir)
    .filter((name) => name.endsWith(".txt"))
    .map((name) => join(touchedDir, name))
    .sort();
}

/** Paths recorded for `sessionId` in the touched-paths log, each a POSIX string
 * relative to repoRoot starting with `.planctl/`. Mirrors
 * invocation._read_touched_paths: a path with `..` or a non-`.planctl/` prefix
 * is a bug upstream and throws loud, never silently dropped. */
function readTouchedPaths(repoRoot: string, sessionId: string): string[] {
  const touchedDir = join(
    repoRoot,
    ".planctl",
    "state",
    "sessions",
    sessionId,
    "touched",
  );
  if (!existsSync(touchedDir)) {
    return [];
  }

  const paths: string[] = [];
  for (const name of readdirSync(touchedDir)) {
    if (!name.endsWith(".txt")) {
      continue;
    }
    const file = join(touchedDir, name);
    const raw = readFileSync(file, "utf-8").trim();
    if (!raw) {
      continue;
    }
    // Security: reject traversal and non-.planctl/ paths.
    if (raw.split("/").includes("..")) {
      throw new Error(
        `Touched-paths record contains path traversal: '${raw}' ` +
          `(file: ${file}). This is a bug — report it.`,
      );
    }
    if (!raw.startsWith(".planctl/")) {
      throw new Error(
        `Touched-paths record contains non-.planctl/ path: '${raw}' ` +
          `(file: ${file}). This is a bug — report it.`,
      );
    }
    paths.push(raw);
  }
  return paths;
}

/** The set of dirty (modified/untracked) .planctl/ paths from git. Mirrors
 * invocation._dirty_planctl_paths: `git status --porcelain --untracked-files=all
 * -- .planctl/` (the flag is load-bearing — without it new files show as a
 * directory-level `?? .planctl/epics/` and the intersection returns empty). The
 * parse matches Python exactly: line[3:].strip(), rename "a -> b" takes b. */
function dirtyPlanctlPaths(repoRoot: string): Set<string> {
  const proc = Bun.spawnSync(
    [
      "git",
      "status",
      "--porcelain",
      "--untracked-files=all",
      "--",
      ".planctl/",
    ],
    { cwd: repoRoot },
  );
  const paths = new Set<string>();
  for (const line of proc.stdout.toString().split("\n")) {
    if (line.length < 4) {
      continue;
    }
    // git status --porcelain: XY <path> (first 3 chars are status + space).
    let rel = line.slice(3).trim();
    // Handle renames: "old -> new" — take the new path.
    if (rel.includes(" -> ")) {
      rel = rel.split(" -> ", 2)[1] as string;
    }
    if (rel) {
      paths.add(rel);
    }
  }
  return paths;
}
